Find more records in PrimoCentral

olli-gold opened the pull request:
By default a query in PrimoCentral only returns records, which are marked as "fulltext available" in the holdingsfile, which is used by PrimoCentral. This can get changed using a simple parameter in the PrimoCentral query URL.

This PR implements this parameter by adding a checkbox facet, which can be used to drop that default filter.

olli-gold commented:
OK, I guess I have to do a little more work to complete this PR, since it conflicts with the PrimoCentral-Backend tests.

olli-gold commented:
OK, its done, unless anybody else sees something I have been missing.

demiankatz commented:
A couple of things:

1.) I see you're using a value in the filter list to maintain this new parameter, but I don't see code that is actually removing that special value from the filter list. Am I missing something, or is this "fake" value getting passed through as a filter? If it is getting passed through, I assume it's not having any adverse effects right now, but should we remove it to ensure it doesn't cause side effects down the road?

2.) We have an existing string in our translation files used for the equivalent functionality in Summon: `
add_other_libraries = "Include articles at other libraries"` -- perhaps this would be a good option instead of the "Show more" you are using in the .ini file... but maybe the language isn't quite appropriate for Primo. I'll leave it to your discretion!

Thanks for sharing this!

olli-gold commented:
Thanks for your comments, Demian.

I also thought about removing the special value, but as far as I see, this would mean, that the special facet will not be marked as checked (and I want it to be marked as checked, when I click on it). For me, this additional facet value had no side effects, but I also cannot assure, that this applies to any configuration. But I will verify and test that again.

About your second comment: Its hard to find a good label for this facet, but I think "Include articles at other libraries" does not match this function at all. For instance, PrimoCentral also offers content from Wikipedia (if you want to include stuff like that in your discovery system...) or other free resources, which are not associated to any library at all. So that label could cause confusion. But I guess Summon and Primo Central will not be getting used together in one discovery system, and so we could really reuse the variable add_other_libraries. The label can be always changed in a custom configuration, though.

demiankatz commented:
Regarding the special value, you can remove it without losing the marked checkbox. VuFind maintains the facet values internally, then translates those values into a format that gets sent to the search backend. If you remove the special value inside the backend, that should only affect which values get sent to the Primo API; it shouldn't bleed back out to the VuFind UI side of things. I think it's worth a try, just to be on the safe side!

Regarding the label, if you feel it's inappropriate, it's fine to leave things as they are... though you might want to look through en.ini to see if there is a different string that can be reused. It's always nice to use an existing translated string when possible to make the feature as useful as possible to the whole international audience... but, of course, if there's nothing appropriate there, we can either add something, or just assume that, since this is off by default, users can figure out an appropriate translation themselves when they turn it on.

olli-gold commented:
Right, I had figured out already, that removing the facet from the query string does not affect the checkbox mark. So everything is working fine and its a lot cleaner than before. Thank you for that hint.

I have also changed the default value for the label to the Summon-label. I guess these functions are comparable, and so the translation string should also be treated the same way. For us its not appropriate, but as I mentioned, that's just a matter of configuration and it doesn't hurt to have the existing label in the configuration file by default.

demiankatz commented:
Thanks -- sounds good to me!

A couple more very small things:

1.) I think you can simplify:

<pre>
        if (array_key_exists('filterList', $params)
            && array_key_exists('pcAvailability', $params['filterList'])
        ) {
</pre>

to:

<pre>
        if (isset($params['filterList']['pcAvailability'])) {
</pre>

It's generally preferable to use isset instead of array_key_exists unless you are checking for a key whose value is set to null (which requires you to use array_key_exists, because isset treats a value set to null as being unset). But in this situation, I think it will be faster and more readable to use isset.

2.) I see that you've copied and pasted the `edition:1st* = "Edition"` contrived hypothetical example from the Solr configuration into the Primo configuration. I don't think that's a good example here, because I doubt that query will actually work in Primo. I'd recommend either eliminating it or replacing it with a different sample checkbox query that makes more sense in Primo. (Obviously, a working example would be ideal, but I don't want to make you waste time thinking up an example if nothing obvious springs to mind immediately).

I'm happy to take care of these things myself if you don't have time to deal with them -- but wanted to share my final thoughts before taking matters into my own hands.

olli-gold commented:
Thanks for your additional hints. I have changed the code as you suggested and I have chosen the simple way for comment number 2 and just removed the contrived example.

Thanks again!

demiankatz commented:
Great. I've merged this to master, collapsing the commit history and adjusting some minor style points to reduce the line count. Here's the final result: https://github.com/vufind-org/vufind/commit/d7e18f6ce9917c8afa5655b460daa4e239b7dd7b

Unfortunately, I can't test this at the moment. If you could pull the latest master and confirm that this is still working, that would be a big help! Thanks again for sharing!

olli-gold commented:
I have pulled the latest master and tested it and can confirm that its working (at least for me, but I think, it should also be globally ok).

demiankatz commented:
Thanks for the double-check!
